EIPS/eip-2200.md

```diff
@@ -68,29 +98,221 @@ the following logic:
         refund counter.
       * Otherwise, add `SSTORE_RESET_GAS - SLOAD_GAS` gas to refund
         counter.
-* If *gasleft* is less than or equal to 2300, fail the current call
-  frame with 'out of gas' exception.
+* If *gasleft* is less than or equal to gas stipend, fail the current
+  call frame with 'out of gas' exception.
 
-An implementation should also note EIP-1283's refund counter
-implementation details, in the *Specification* section.
+An implementation should also note that with the above definition, if
+the implementation uses call-frame refund counter, the counter can go
+negative. If the implementation uses transaction-wise refund counter,
+the counter always stays positive.
 
 ## Rationale
 
-The same as EIP-1283's rationale. If EIP-1884 is applied, then `SLOAD_GAS`
-related to SSTORE opcode will also need to be changed.
+This EIP mostly achieves what a transient storage tries to do
+(EIP-1087 and EIP-1153), but without the complexity of introducing the
+concept of "dirty maps", or an extra storage struct.
+
+* We don't suffer from the optimization limitation of
+  EIP-1087. EIP-1087 requires keeping a dirty map for storage changes,
+  and implicitly makes the assumption that a transaction's storage
+  changes are committed to the storage trie at the end of a
+  transaction. This works well for some implementations, but not for
+  others. After EIP-658, an efficient storage cache implementation
+  would probably use an in-memory trie (without RLP encoding/decoding)
+  or other immutable data structures to keep track of storage changes,
+  and only commit changes at the end of a block. For them, it is
+  possible to know a storage's original value and current value, but
+  it is not possible to iterate over all storage changes without
+  incurring additional memory or processing costs.
+* It never costs more gas compared with the current scheme.
+* It covers all usages for a transient storage. Clients that are easy
+  to implement EIP-1087 will also be easy to implement this
+  specification. Some other clients might require a little bit extra
+  refactoring on this. Nonetheless, no extra memory or processing cost
+  is needed on runtime.
+
+Regarding SSTORE gas cost and refunds, see Appendix for proofs of
+properties that this EIP satisfies.
+
+* For *absolute gas used* (that is, actual *gas used* minus *refund*),
+  this EIP is equivalent to EIP-1087 for all cases.
+* For one particular case, where a storage slot is changed, reset to
+  its original value, and then changed again, EIP-1283 would move more
+  gases to refund counter compared with EIP-1087.
+
+Examine examples provided in EIP-1087's Motivation:
+
+* If a contract with empty storage sets slot 0 to 1, then back to 0,
+  it will be charged `20000 + 200 - 19800 = 400` gas.
```

Comment:
Numbers here also should be updated for new `SLOAD_GAS`

Reply:
I added note in proof that we do it under the assumption that `SLOAD_GAS` is `200`.